src/util/timer.rs

```diff
@@ -0,0 +1,195 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::{Ord, Ordering};
+use std::time::Duration;
+use std::collections::BinaryHeap;
+
+use util::time::Instant;
+
+pub struct Timer<T> {
+    pending: BinaryHeap<TimeoutTask<T>>,
+}
+
+impl<T> Timer<T> {
+    pub fn new(capacity: usize) -> Self {
+        Timer {
+            pending: BinaryHeap::with_capacity(capacity),
+        }
+    }
+
+    /// Add a periodic task into the `Timer`.
+    pub fn add_task(&mut self, timeout: Duration, task: T) {
+        let task = TimeoutTask {
+            next_tick: Instant::now() + timeout,
+            task: task,
+        };
+        self.pending.push(task);
+    }
+
+    /// Get the next `timeout` from the timer.
+    pub fn next_timeout(&mut self) -> Option<Instant> {
+        self.pending.peek().map(|task| task.next_tick)
+    }
+
+    /// Pop a `TimeoutTask` from the `Timer`, which should be tick before `instant`.
+    /// If there is no tasks should be ticked any more, None will be returned.
+    ///
+    /// The normal use case is keeping `pop_task_before` until get `None` in order
+    /// to retreive all avaliable events.
+    pub fn pop_task_before(&mut self, instant: Instant) -> Option<T> {
+        if self.pending
+            .peek()
+            .map_or(false, |t| t.next_tick <= instant)
+        {
+            return self.pending.pop().map(|t| t.task);
+        }
+        None
+    }
+}
+
+#[derive(Debug)]
+struct TimeoutTask<T> {
+    next_tick: Instant,
+    task: T,
+}
+
+impl<T> PartialEq for TimeoutTask<T> {
+    fn eq(&self, other: &TimeoutTask<T>) -> bool {
+        self.next_tick == other.next_tick
+    }
+}
+
+impl<T> Eq for TimeoutTask<T> {}
+
+impl<T> PartialOrd for TimeoutTask<T> {
+    fn partial_cmp(&self, other: &TimeoutTask<T>) -> Option<Ordering> {
+        other.next_tick.partial_cmp(&self.next_tick)
+    }
+}
+
+impl<T> Ord for TimeoutTask<T> {
+    fn cmp(&self, other: &TimeoutTask<T>) -> Ordering {
+        self.partial_cmp(other).unwrap_or(Ordering::Equal)
```

Comment:
It seems we can't choose `Ordering::Equal` arbitrarily, it can not give a stable order when we mix `Instant::now` and `Instant::now_coarse`, sorry! :slightly_frowning_face:

Maybe compare with inner `Timespec`s, it implements `Ord` correctly.  Please add caveats about the behavior of comparing with different variants of the `Instant`.

----

Example:

```rust
    #[test]
    fn test_order_timer() {
        let t1 = TimeoutTask {
            next_tick: Instant::now(),
            task: 1,
        };
        let t2 = TimeoutTask {
            next_tick: Instant::now_coarse(),
            task: 2,
        };
        let t3 = TimeoutTask {
            next_tick: Instant::now(),
            task: 3,
        };
        {
            let mut array = Vec::new();
            array.push(t1.clone());
            array.push(t2.clone());
            array.push(t3.clone());
            array.sort();
            println!("{:?}", array);
        }
        {
            let mut array = Vec::new();
            array.push(t1.clone());
            array.push(t3.clone());
            array.push(t2.clone());
            array.sort();
            println!("{:?}", array);
        }
        {
            let mut array = Vec::new();
            array.push(t3.clone());
            array.push(t2.clone());
            array.push(t1.clone());
            array.sort();
            println!("{:?}", array);
        }
    }
```